src/poll.rs

```diff
@@ -564,107 +967,122 @@ impl Drop for RegistrationInner {
 
+    /// Poll the queue for new events
     fn poll(&self, dst: &mut sys::Events) {
-        let ready = self.take_ready();
-
-        // TODO: Cap number of nodes processed
-        for node in ready {
-            let mut events;
-            let opts;
-
-            {
-                let node_ref = node.as_ref().unwrap();
-                opts = node_ref.poll_opts();
-
-                // Atomically read queued. Use Acquire ordering to set a
-                // barrier before reading events, which will be read using
-                // `Relaxed` ordering. Reading events w/ `Relaxed` is OK thanks to
-                // the acquire / release hand off on `queued`.
-                let mut queued = node_ref.queued.load(Ordering::Acquire);
-                events = node_ref.poll_events();
-
-                // Enter a loop attempting to unset the "queued" bit or requeuing
-                // the node.
-                loop {
-                    // In the following conditions, the registration is removed from
-                    // the readiness queue:
-                    //
-                    // - The registration is edge triggered.
-                    // - The event set contains no events
-                    // - There is a requested delay that has not already expired.
-                    //
-                    // If the drop flag is set though, the node is never queued
-                    // again.
-                    if event::is_drop(events) {
-                        // dropped nodes are always processed immediately. There is
-                        // also no need to unset the queued bit as the node should
-                        // not change anymore.
-                        break;
-                    } else if opts.is_edge() || event::is_empty(events) {
-                        // An acquire barrier is set in order to re-read the
-                        // `events field. `Release` is not needed as we have not
-                        // mutated any field that we need to expose to the producer
-                        // thread.
-                        let next = node_ref.queued.compare_and_swap(queued, 0, Ordering::Acquire);
-
-                        // Re-read in order to ensure we have the latest value
-                        // after having marked the registration has dequeued from
-                        // the readiness queue. Again, `Relaxed` is OK since we set
-                        // the barrier above.
-                        events = node_ref.poll_events();
-
-                        if queued == next {
-                            break;
-                        }
+        // `until` is set with the first node that gets re-enqueued due to being
+        // set to have level-triggered notifications. This prevents an infinite
+        // loop where `Poll::poll` will keep dequeuing nodes it enqueues.
+        let mut until = ptr::null_mut();
+
+        'outer:
+        while dst.len() < dst.capacity() {
```

Comment:
Yes, definitely a starvation potential. Another option would be to alternate which queue gets polled first. So, first call to `poll` would poll the selector, then the readiness queue, the second would poll the readiness queue then the selector...